scripts/run_inference.py

```diff
@@ -178,14 +181,31 @@ def load(
         try:
             with AdapterConfigPatcher(checkpoint_path, overrides):
                 try:
-                    model = AutoPeftModelForCausalLM.from_pretrained(
+                    if base_model_name_or_path is None:
+                        raise KeyError("base_model_name_or_path has to be passed")
+                    modelbase = AutoModelForCausalLM.from_pretrained(
+                        base_model_name_or_path,
```

Comment:
its not very clear why cant you load the PEFT model first, then do a resize of the embedding layer. 

Also is there a use-case to do this for inference? Since `fms-hf-trainer` does not aim to support inference, it may be better to only do this for training, since the main concern is training throughput.

Reply:
> its not very clear why cant you load the PEFT model first, then do a resize of the embedding layer.

I see there are couple of ways to load PEFT model, but let me answer the problem with the usage of `AutoPeftModelForCausalLM` which is what being replaced in this PR. 

The base model is being internally resized by the PEFT library (https://github.com/huggingface/peft/blob/018a1f49c4a6e500c243efac61d72341879ea9f6/src/peft/auto.py#L128) based on the tokenizer, however, in our case, the tokenizer would not represent the true size of the embedding layer since it does not account for additional increase that was made to match the model embedding layer size to the nearest given multiple. 

Also, the current approach in this PR has been also recommended in their docs - https://huggingface.co/docs/peft/v0.8.2/en/developer_guides/troubleshooting#extending-the-vocabulary 

from docs
```
For inference, load the base model first and resize it the same way you did before you trained the model. After you’ve resized the base model, you can load the PEFT checkpoint.
```

Please let me know if there is a better way to handle this, thanks.


> Also is there a use-case to do this for inference? Since fms-hf-trainer does not aim to support inference, it may be better to only do this for training, since the main concern is training throughput.

@fabianlim The motivation behind these changes was to pass the unit tests around PEFT.

Reply:
thanks, while this script is not used in any product, we do use it in platform to quickly test changes in tuning with HF libraries instead of vLLLM deployment each time. So good to keep the script functional. Thank you for the change

Reply:
@kmehant could you just document here which test was the failing one?

Reply:
sure @fabianlim

Reply:
@fabianlim @Ssukriti 
Tests that were failing.

1. test_run_causallm_pt_and_inference
2. test_run_causallm_pt_and_inference_with_formatting_data
3. test_run_causallm_lora_and_inference[default]
4. test_run_causallm_lora_and_inference[custom_target_modules]
5. test_run_causallm_lora_and_inference[all_linear_target_modules]

Reply:
@kmehant so we are clear on why the above tests are failing. Since resizing embedding table does not change model output, so why do they fail?

Reply:
@fabianlim 
The reasoning is mentioned here - https://github.com/foundation-model-stack/fms-hf-tuning/pull/227#discussion_r1662692069

Here is the summary
```
FAILED tests/test_sft_trainer.py::test_run_causallm_pt_and_inference - RuntimeError: Error(s) in loading state_dict for PeftModelForCausalLM:
FAILED tests/test_sft_trainer.py::test_run_causallm_pt_and_inference_with_formatting_data - RuntimeError: Error(s) in loading state_dict for PeftModelForCausalLM:
FAILED tests/test_sft_trainer.py::test_run_causallm_pt_and_inference_JSON_file_formatter - RuntimeError: Error(s) in loading state_dict for PeftModelForCausalLM:
FAILED tests/test_sft_trainer.py::test_run_causallm_lora_and_inference[default] - RuntimeError: Error(s) in loading state_dict for PeftModelForCausalLM:
FAILED tests/test_sft_trainer.py::test_run_causallm_lora_and_inference[custom_target_modules] - RuntimeError: Error(s) in loading state_dict for PeftModelForCausalLM:
FAILED tests/test_sft_trainer.py::test_run_causallm_lora_and_inference[all_linear_target_modules] - RuntimeError: Error(s) in loading state_dict for PeftModelForCausalLM:
```

One trace for your reference - 
```
E           RuntimeError: Error(s) in loading state_dict for PeftModelForCausalLM:
E           	size mismatch for base_model.model.model.embed_tokens.weight: copying a param with shape torch.Size([32008, 64]) from checkpoint, the shape in current model is torch.Size([32001, 64]).
E           	size mismatch for base_model.model.lm_head.weight: copying a param with shape torch.Size([32008, 64]) from checkpoint, the shape in current model is torch.Size([32001, 64]).

.tox/py/lib/python3.9/site-packages/torch/nn/modules/module.py:2189: RuntimeError
```